File: client/Assets/LuaSupport/Cell/LuaPB.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using google.protobuf;

public class LuaPB
{
    public static string GetTestData( )
    {

       // var s = Encoding.Default.GetString(stream.ToArray());        

        return default(string);
    }

    public static void TestStream( PBStream s )
    {
        var p = new tutorial.Person();
        p.name = "hello";
        p.test.Add(1);
        p.test.Add(2);
        p.phone.Add(new tutorial.PhoneNumber
        {
            number = "789",
            type = tutorial.PhoneType.WORK,
        });

        p.phone.Add(new tutorial.PhoneNumber
        {
            number = "456",

            type = tutorial.PhoneType.HOME,
        });

        var stream = new MemoryStream();
        ProtoBuf.Serializer.Serialize(stream, p);

        s.Stream.Position = 0;
        var data = ProtoBuf.Serializer.Deserialize<tutorial.Person>(s.Stream);
        var compareStream = new MemoryStream();
        ProtoBuf.Serializer.Serialize(compareStream, data);


        if (CompareMemoryStreams(stream, compareStream))
        {
            int a = 1;
        }
    }

    private static bool CompareMemoryStreams(MemoryStream ms1, MemoryStream ms2)
    {
        if (ms1.Length != ms2.Length)
            return false;
        ms1.Position = 0;
        ms2.Position = 0;

        var msArray1 = ms1.ToArray();
        var msArray2 = ms2.ToArray();

        return msArray1.SequenceEqual(msArray2);
    }

    static DescriptorPool _pool = new DescriptorPool();

    public static void RegisterFile( string filename )
    {        
        using( FileStream reader = File.OpenRead(filename) )
        {
            var data = ProtoBuf.Serializer.Deserialize<FileDescriptorSet>(reader);

            reader.Dispose();

            _pool.Init(data);
        }

        var msg = _pool.GetMessage("tutorial.Person");

    }

    public static DescriptorPool GetPool( )
    {
        return _pool;
    }


    public static int TagSize( int fieldNumber )
    {
        return VarintSize32((uint)(fieldNumber << kTagTypeBits));
    }


    const int kTagTypeBits = 3;
    public static int VarintSize32( uint value )
    {
        if (value < (1 << 7))
        {
            return 1;
        }
        else if (value < (1 << 14))
        {
            return 2;
        }
        else if (value < (1 << 21))
        {
            return 3;
        }
        else if (value < (1 << 28))
        {
            return 4;
        }

        return 5;
    }

    public static int Int32Size( int value )
    {
        if (value < 0)
            return 10;

        return VarintSize32((uint)value);
    }    
}

File: client/Assets/LuaSupport/Cell/LuaPB.lua
local FieldType_Message = 11

local fieldSizeMapper = {

	-- float
	[2] = 4,

	-- int32
	[5] = LuaPB.Int32Size,

	-- bool
	[8] = 1,

	-- string
	[9] = function( str )
		local len = #str

		return LuaPB.VarintSize32( len ) + len
	end,
}


local function FieldSize( fd, value )

	local v = fieldSizeMapper[fd.Type]
	local typeOfV = type(v)
	if typeOfV == "function" then
		return v(value)
	end

	return v
end


local function RawByteSize( msgD, t )

	local size = 0

	for k, v in pairs( t ) do
	
		local fd = msgD:GetFieldByName( k )
		
		if fd ~= nil then
		
			if fd.Type == FieldType_Message then
			
				if fd.IsRepeated then
			
					for listK, listV in ipairs(v) do
					
						size = size + LuaPB.TagSize( fd.Number )
						
						local structSize = RawByteSize( fd.MessageType, listV )
						size = size + LuaPB.VarintSize32( structSize ) + structSize
					
					end
					
				else
				
					size = size + LuaPB.TagSize( fd.Number )
					size = size + RawByteSize( fd.MessageType, v )
				end
			
			else 
				-- 普通值
				
				if fd.IsRepeated then
				
					for listK, listV in ipairs(v) do
						size = size + LuaPB.TagSize( fd.Number )
						size = size + FieldSize( fd, listV )
					end
					
				else
				
					size = size + LuaPB.TagSize( fd.Number )
					size = size + FieldSize( fd, v )
				
				end
			
			end 
		
		
		end -- if fd ~= nil then
		
	end
	
	return size


end


local function RawEncode( msgD, t )


	for k, v in pairs( t ) do
	
		local fd = msgD:GetFieldByName( k )
		
		if fd ~= nil then
		
			if fd.Type == FieldType_Message then
			
				if fd.IsRepeated then
			
					for listK, listV in ipairs(v) do
					
					
					end
					
				else
				
				end
			
			else 
				-- 普通值
				
				if fd.IsRepeated then
				
					for listK, listV in ipairs(v) do

					end
					
				else
				
					
				
				end
			
			end 
		
		
		end -- if fd ~= nil then
		
	end

end



function luapb_bytesize( name,  t )

	local pool = LuaPB.GetPool()
	
	local msgD = pool:GetMessage( name )

	return RawByteSize( msgD, t )
end


function luapb_encode( name, t )

	local pool = LuaPB.GetPool()
	
	local msgD = pool:GetMessage( name )

	
end

